akka-typed/src/main/scala/akka/typed/Extensions.scala

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed
+
+import akka.annotation.DoNotInherit
+
+/**
+ * Marker trait/interface for extensions. An extension can be registered in the ActorSystem and is guaranteed to only
+ * have one instance per [[ActorSystem]] instance per [[ExtensionId]]. The extension internals must be thread safe.
+ * For mutable state it should be preferred to use an `Actor` rather than extensions as first choice.
+ *
+ * @see [[ExtensionId]]
+ */
+trait Extension
+
+/**
+ * Identifier and factory for an extension. Is used to look up an extension from the `ActorSystem`, and possibly create
+ * an instance if no instance was already registered.
+ *
+ * Scala API:
+ *
+ * {{{
+ * object MyExt extends ExtensionId[Ext] {
+ *
+ *   override def createExtension(system: ActorSystem[_]): MyExt = new MyExt(system)
+ *
+ *   // Java API: retrieve the extension for the given system.
+ *   override def get(system: ActorSystem[_]): MyExt = super.get(system)
+ * }
+ *
+ * class MyExt(system: ActorSystem[_]) extends Extension {
+ *   ...
+ * }
+ * }}}
+ *
+ * Java API:
+ *
+ * {{{
+ *
+ * public class MyExt extends AbstractExtensionId<MyExtImpl> {
+ *   // single instance of the identifier
+ *   private final static MyExt instance = new MyExt();
+ *
+ *   // protect against other instances than the singleton
+ *   private MyExt() {}
+ *
+ *   // This static method singleton accessor is needed to be able to enable the extension through config when
+ *   // implementing extensions in Java.
+ *   public static MyExt getInstance() {
```

Comment:
ah, I see in [docs](http://doc.akka.io/docs/akka/2.5/java/extending-akka.html) that it was not very intuitive from java. I think this convention is good. Document is even more clear that the method must `static` and named `getInstance`. Write that also outside of this code example.